add server binary tutorial (#282)
temporalio/temporal-learning: docs/tutorials/infrastructure/binary/index.md

If you need a deployment that fits in between these options -- for example, if you need to scale for multiple users, with fine-grained control over your deployment parameters, but without the overhead of Kubernetes -- you can deploy a Temporal Service using the official server binaries.

In this tutorial, you'll learn how to configure and deploy the two binaries needed for a complete Temporal Service (the core server and the UI server). You'll create systemd unit files to gracefully run and restart the Temporal Service automatically upon server startup, and you'll deploy an Nginx reverse proxy to handle web traffic ingress. This will give you everything you need to run a production Temporal Service, and evaluate how to scale further or [migrate to Temporal Cloud](https://temporal.io/cloud). Let's get started.

## Prerequisites

> - A Linux server with SSH access and the Nginx web server installed. This can be a new Ubuntu server instance with no additional configuration performed. However, you will need a version of Nginx built with HTTP/2 support (at least version 1.25.1), which may not be available by default in some environments. On Ubuntu, you can use [this PPA](https://launchpad.net/~ondrej/+archive/ubuntu/nginx) to install a compatible Nginx.

Reviewer: ```suggestion
- A Linux server with SSH access and the Nginx web server installed. This can be a new Ubuntu server instance with no additional configuration performed. However, you will need a version of Nginx built with HTTP/2 support (at least version 1.25.1), which may not be available by default in some environments. On Ubuntu, you can use [this PPA](https://launchpad.net/~ondrej/+archive/ubuntu/nginx) to install a compatible Nginx.

Run the following commands to install the latest version of `nginx`:
```bash
sudo add-apt-repository ppa:ondrej/nginx
sudo apt update
sudo apt install -y nginx
```

```

```bash
sudo mv ui-server /usr/bin/temporal-ui-server
sudo chmod +x /usr/bin/temporal-ui-server
```

At this point, you've downloaded everything you need. The last thing to do is create a `temporal` user on your server that has the appropriate permissions to run the Temporal Service, and directory accessible to this user to store your data in. Run the following commands:

```bash
sudo useradd temporal
sudo mkdir /etc/temporal
sudo chown temporal /etc/temporal
```

Next, you'll create configuration files for both the Temporal Server and the UI Server in the `/etc` directory.

## Configuring the Temporal Binaries

In this tutorial, you'll configure a connection to a SQLite database, since it doesn't require any additional dependencies. Using your favorite text editor, open a new file called `/etc/temporal-server.yaml`:

```bash
sudo vim /etc/temporal-server.yaml
```

Paste the following contents into the file for a starting configuration. You can update any of these values later.

```
log:
  stdout: true
  level: info

persistence:
  defaultStore: sqlite-default
  visibilityStore: sqlite-visibility

>   numHistoryShards: 1

Reviewer 2: dev-server uses 4 hard-coded here. I think it may be a bit academic because SQLite doesn't allow concurrent writes, but I wouldn't take it lower than 4 anyway. A real production system would generally not have this lower than 256, probably 512.

  datastores:
    sqlite-default:
      sql:
        pluginName: "sqlite"
        databaseName: "/etc/temporal/default.db"
        connectAddr: "localhost"
        connectProtocol: "tcp"
        connectAttributes:
          cache: "private"
          setup: true

    sqlite-visibility:
      sql:
        pluginName: "sqlite"
        databaseName: "/etc/temporal/visibility.db"
        connectAddr: "localhost"
        connectProtocol: "tcp"
        connectAttributes:
          cache: "private"
          setup: true

global:
  membership:
    maxJoinDuration: 30s
    broadcastAddress: "127.0.0.1"
  pprof:
    port: 7936

services:
  frontend:
    rpc:
      grpcPort: 7233
      membershipPort: 6933
      bindOnLocalHost: true
      httpPort: 7243

  matching:
    rpc:
      grpcPort: 7235
      membershipPort: 6935
      bindOnLocalHost: true

  history:
    rpc:
      grpcPort: 7234
      membershipPort: 6934
      bindOnLocalHost: true

  worker:
    rpc:

>       grpcPort: 7239

Reviewer 2: `grpcPort` and `bindOnLocalHost` can go for `worker`, it doesn't listen as it doesn't have it's own service. These lines are still present in our defaults/samples just because they haven't been cleaned up yet,

      membershipPort: 6939
      bindOnLocalHost: true

clusterMetadata:
  enableGlobalNamespace: false
  failoverVersionIncrement: 10
  masterClusterName: "active"
  currentClusterName: "active"
  clusterInformation:
    active:
      enabled: true
      initialFailoverVersion: 1
      rpcName: "frontend"
      rpcAddress: "localhost:7233"
      httpAddress: "localhost:7243"

dcRedirectionPolicy:
  policy: "noop"
```

Save and close the file. Next, you'll create the configuration file for the UI Server. Using your favorite text editor, open a new file called `/etc/temporal-ui-server.yaml`:

```bash
sudo vim /etc/temporal-ui-server.yaml
```

Paste the following contents into the file.

```
temporalGrpcAddress: 127.0.0.1:7233
port: 8233
enableUi: true
cors:
  allowOrigins:

>     - http://localhost:8233

Reviewer 2: This needs to be whatever their custom domain is rather than localhost, otherwise things like Terminate etc won't work as they will fail CORS checks.

   Main PID: 19925 (temporal-server)
      Tasks: 22 (limit: 18707)
     Memory: 62.7M
     CGroup: /system.slice/temporal.service
             └─19925 temporal-server -c /etc -e temporal-server start

Jul 08 11:24:42 Omelas temporal-server[19925]: {"level":"info","ts":"2024->
Jul 08 11:24:42 Omelas temporal-server[19925]: {"level":"info","ts":"2024->
Jul 08 11:24:42 Omelas temporal-server[19925]: {"level":"info","ts":"2024-
```

Next, repeat these steps for the UI server. Open a new file called `/etc/systemd/system/temporal-ui.service`:

```bash
sudo vim /etc/systemd/system/temporal-ui.service
```

Add the following contents:

```
[Unit]
Description=Temporal UI Server
After=network.target

[Service]
User=temporal
Group=temporal
ExecStart=temporal-ui-server -c /etc -e temporal-ui-server start

[Install]
WantedBy=multi-user.target
```

Save and close the file, then `start` the UI Server service, and `enable` it to run on boot automatically:

```bash
sudo systemctl start temporal-ui
sudo systemctl enable temporal-ui
```

Use systemctl status temporal to verify that it started successfully:

```bash
sudo systemctl status temporal-ui
```

> Both services should now be running in the background. 

Author: I forget if this was in the actual DO style guide or not, but some folks pushed for "receive" instead of "see" on the basis that it was potentially exclusionary for folks using screen readers. I personally feel this is a pretty challenging and potentially unnecessary style point, but raising it here anyway.

        proxy_read_timeout 300;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Real-PORT $remote_port;
        allow your_ip_addresses;
        deny all;
    }

    listen 443 ssl;
    # RSA certificate
    ssl_certificate /etc/letsencrypt/live/your_domain/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/your_domain/privkey.pem;

    # Redirect non-https traffic to https
    if ($scheme != "https") {
        return 301 https://$host$request_uri;
    }
}
```

You can read this configuration as having three main “blocks” to it. The first block, coming before the `location /` line, contains a boilerplate Nginx configuration for serving a website on the default HTTP port, 80. The `location /` block contains a configuration for proxying incoming connections to the Temporal Web UI, running on port 8233 internally, while preserving SSL. The configuration at the end of the file, after the `location /` block, loads your LetsEncrypt SSL keypairs and redirects HTTP connections to HTTPS.

> Note the `allow your_ip_address;` line. You should replace this with the IP address that you'll need to access the Temporal Web UI from. You can add additional addresses in the same range using [CIDR notation](https://www.digitalocean.com/community/tutorials/understanding-ip-addresses-subnets-and-cidr-notation-for-networking), or add additional consecutive `allow` lines for multiple IPs. The `deny all;` on the following line will block traffic to all but the specified IPs. Because Temporal does not use any kind of authentication by default, restricting traffic by IP address is the least complex way of providing secure access.

Author: Yeah, I thought about adding a whole firewall step, but I'm reluctant to do so because we'd have to be somewhat opinionated about the choice of firewall, and it's technically an unnecessary dependency when Nginx can do this.

> Save and close the file. Next, you’ll need to activate this new configuration. Nginx’s convention is to create symbolic links (like shortcuts) from files in `sites-available/` to another folder called `sites-enabled/` as you decide to enable or disable them. Using full paths for clarity, make that link:

Author: I had one and took it out because I was concerned it was too DigitalOcean-y to go through that many steps. Happy to re-add!

```bash
sudo ln -s /etc/nginx/sites-available/temporal-ui /etc/nginx/sites-enabled/temporal-ui
```

By default, Nginx includes another configuration file at `/etc/nginx/sites-available/default`, linked to `/etc/nginx/sites-enabled/default`, which also serves its default index page. You’ll need to disable that rule by removing it from `/sites-enabled`, because it conflicts with your new Temporal configuration:

```bash
sudo rm /etc/nginx/sites-enabled/default
```

Before reloading Nginx with this new configuration, you'll create another reverse proxy for gRPC API connections to Temporal itself. Create another Nginx configuration at `/etc/nginx/sites-available/temporal`:

```bash
sudo vim /etc/nginx/sites-available/temporal
```

Paste the following into the new configuration file, being sure to replace `your_domain` with your domain name.

```
server {
    listen 7233;
    listen [::]:7233;
    server_name your_domain

> 

Reviewer: Do you mean to add access and error log to these?

    http2 on;

    location / {
        grpc_pass localhost:7233;
        allow your_client_ip_address;
        deny all;
    }
}
```

This configuration is shorter than the previous one, because Nginx only needs to use the `grpc_pass` directive to send gRPC traffic to the server using the same external port (7233) that Temporal's gRPC API runs on. Again, don't forget the the `allow your_client_ip_address;` line. In this case, you'll need an `allow` statement or IP range for everywhere that you plan to run your Temporal Workers, or any other Temporal Client, or connect via the `temporal` CLI.

Save and close the file, and create a symbolic link as before:

```bash
sudo ln -s /etc/nginx/sites-available/temporal /etc/nginx/sites-enabled/temporal
```

Now you can restart your Nginx service, so it will reflect your new configuration:

```bash
sudo systemctl restart nginx
```

Navigate to your_domain in a web browser, and you should receive the Temporal Web UI. At this point, you're finished with configuration. In the final step, you'll review the logs generated by your Temporal Service, as well as your options for connecting to it from the Temporal CLI or SDK.

> ## Logging, Restarting, and Connecting to the Temporal Service

Reviewer: Since each of these points are _super_ short, it feels odd to distinctly call them out. If they were h3s with more content I think it would be more appropriate, but since they're a sentence I think Interacting summarizes them better. 

```suggestion
## Interacting with the Temporal Service
```

> In the last step, when you configured your Nginx reverse proxy, you added logging to `access_log /var/log/nginx/temporal.access.log;`. You can check this file if you ever need to review access logs to the Temporal Web UI.

Reviewer: ```suggestion
In the last step, when you configured your Nginx reverse proxy, you enabled logging to `/var/log/nginx/temporal.access.log;` when you set the `access_log` parameter. Check this file if you ever need to review access logs to the Temporal Web UI.
```

In this tutorial, you'll learn how to configure and deploy the two binaries needed for a complete Temporal Service (the core server and the UI server). You'll create `systemd` unit files to gracefully run and restart the Temporal Service automatically upon server startup, and you'll deploy an Nginx reverse proxy to handle web traffic ingress. This will give you everything you need to run a production Temporal Service, and evaluate how to scale further or [migrate to Temporal Cloud](https://temporal.io/cloud). Let's get started.

## Prerequisites

- A Linux server with SSH access and the Nginx web server installed. This can be a new Ubuntu server instance with no additional configuration performed. However, you will need a version of Nginx built with HTTP/2 support (at least version 1.25.1), which may not be available by default in some environments. On Ubuntu, you can use [this PPA](https://launchpad.net/~ondrej/+archive/ubuntu/nginx) to install a compatible Nginx.

> - To enable HTTPS in the browser, you will need SSL certificates and your own domain name pointing to the server.

Reviewer: ```suggestion
- To enable HTTPS in the browser, you will need SSL certificates and your own domain name pointing to the server.

Once you have your domain pointing the the server, one option to get an SSL certificate is to use Certbot. This tutorial will assume this was the option you chose. Use the commands below, replacing `YOUR_DOMAIN` with the domain pointing to your server to generate the certificate:

```bash
systemctl stop nginx # ensure that nginx is stopped and port 80 is available
sudo snap install --classic certbot
sudo ln -s /snap/bin/certbot /usr/bin/certbot
sudo certbot certonly --standalone -d YOUR_DOMAIN
```

```

While your Temporal Service is currently running, it is still only available on the internal `localhost` network. Next you should make it available externally and secure connections to it.
When `certbot` retrieves certificates, by default, it stores them in `/etc/letsencrypt/live/your_domain`. Check to make sure that you have them:

```bash
sudo ls /etc/letsencrypt/live/your_domain
```

```
README  cert.pem  chain.pem  fullchain.pem  privkey.pem
```

Now, you can configure an Nginx *reverse proxy* to expose your Temporal Service to external connections. Putting a web server such as Nginx in front of other web-facing applications can improve performance and reduce the complexity of securing a site. Nginx can take care of restricting access and securely handling requests from your clients to Temporal. You'll configure the UI server first, which uses regular HTTP (web) traffic:

Nginx allows you to add per-site configurations to individual files in a subdirectory called `sites-available/`. Using your favorite text editor, create a new Nginx configuration at `/etc/nginx/sites-available/temporal-ui`:

```bash
sudo vim /etc/nginx/sites-available/temporal-ui
```

> Paste the following into the new configuration file, being sure to replace `your_domain` with your domain name.

Reviewer: ```suggestion
Paste the following into the new configuration file, being sure to replace `YOUR_DOMAIN` with your domain name.
```

```
server {
    listen 80 default_server;
    listen [::]:80 default_server;
    server_name your_domain www.your_domain

    access_log /var/log/nginx/temporal.access.log;
    error_log /var/log/nginx/temporal.error.log;

    location / {
        proxy_pass http://127.0.0.1:8233;
        proxy_http_version 1.1;
        proxy_read_timeout 300;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $http_host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Real-PORT $remote_port;

>         allow your_ip_addresses;

Reviewer: I have found that all caps draws the eye to places you need to change. 
```suggestion
        allow YOUR_IP_ADDRESSES;
```

        deny all;
    }

    listen 443 ssl;
    # RSA certificate

>     ssl_certificate /etc/letsencrypt/live/your_domain/fullchain.pem;

Reviewer: ```suggestion
    ssl_certificate /etc/letsencrypt/live/YOUR_DOMAIN/fullchain.pem;
```

>     ssl_certificate_key /etc/letsencrypt/live/your_domain/privkey.pem;

Reviewer: ```suggestion
    ssl_certificate_key /etc/letsencrypt/live/YOUR_DOMAIN/privkey.pem;
```

    # Redirect non-https traffic to https
    if ($scheme != "https") {
        return 301 https://$host$request_uri;
    }
}
```

You can read this configuration as having three main “blocks” to it. The first block, coming before the `location /` line, contains a boilerplate Nginx configuration for serving a website on the default HTTP port, 80. The `location /` block contains a configuration for proxying incoming connections to the Temporal Web UI, running on port 8233 internally, while preserving SSL. The configuration at the end of the file, after the `location /` block, loads your LetsEncrypt SSL keypairs and redirects HTTP connections to HTTPS.

> Note the `allow your_ip_address;` line. You should replace this with the IP address that you'll need to access the Temporal Web UI from. You can add additional addresses in the same range using [CIDR notation](https://www.digitalocean.com/community/tutorials/understanding-ip-addresses-subnets-and-cidr-notation-for-networking), or add additional consecutive `allow` lines for multiple IPs. The `deny all;` on the following line will block traffic to all but the specified IPs. Because Temporal does not use any kind of authentication by default, restricting traffic by IP address is the least complex way of providing secure access.

Reviewer: ```suggestion
Note the `allow YOUR_IP_ADDRESSES;` line. You should replace this with the IP address that you'll need to access the Temporal Web UI from. You can add additional addresses in the same range using [CIDR notation](https://www.digitalocean.com/community/tutorials/understanding-ip-addresses-subnets-and-cidr-notation-for-networking), or add additional consecutive `allow` lines for multiple IPs. The `deny all;` on the following line will block traffic to all but the specified IPs. Because Temporal does not use any kind of authentication by default, restricting traffic by IP address is the least complex way of providing secure access.
```

Save and close the file. Next, you’ll need to activate this new configuration. Nginx’s convention is to create symbolic links (like shortcuts) from files in `sites-available/` to another folder called `sites-enabled/` as you decide to enable or disable them. Using full paths for clarity, make that link:

```bash
sudo ln -s /etc/nginx/sites-available/temporal-ui /etc/nginx/sites-enabled/temporal-ui
```

By default, Nginx includes another configuration file at `/etc/nginx/sites-available/default`, linked to `/etc/nginx/sites-enabled/default`, which also serves its default index page. You’ll need to disable that rule by removing it from `/sites-enabled`, because it conflicts with your new Temporal configuration:

```bash
sudo rm /etc/nginx/sites-enabled/default
```

Before reloading Nginx with this new configuration, you'll create another reverse proxy for gRPC API connections to Temporal itself. Create another Nginx configuration at `/etc/nginx/sites-available/temporal`:

```bash
sudo vim /etc/nginx/sites-available/temporal
```

> Paste the following into the new configuration file, being sure to replace `your_domain` with your domain name.

Reviewer: ```suggestion
Paste the following into the new configuration file, being sure to replace `YOUR_DOMAIN` with your domain name.
```

```
server {
    listen 7233;
    listen [::]:7233;

>     server_name your_domain

Reviewer: ```suggestion
    server_name YOUR_DOMAIN
```
